src/Eshopworld.Messaging/Messenger.cs

```diff
@@ -1,145 +1,196 @@
+    {
+        internal readonly object Gate = new object();
+        internal readonly string ConnectionString;
+        internal readonly string SubscriptionId;
+
+        internal readonly ISubject<object> MessagesIn = new Subject<object>();
+
+        internal Dictionary<Type, IDisposable> MessageSubs = new Dictionary<Type, IDisposable>();
+        internal Dictionary<Type, ServiceBusAdapter> ServiceBusAdapters = new Dictionary<Type, ServiceBusAdapter>();
+
+        internal ServiceBusAdapter<T> GetQueueAdapterIfExists<T>() where T : class =>
+            ServiceBusAdapters.TryGetValue(typeof(T), out var result)
+                ? (ServiceBusAdapter<T>)result
+                : throw new InvalidOperationException($"Messages/events of type {typeof(T).FullName} haven't been setup properly yet");
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="Messenger"/>.
+        /// </summary>
+        /// <param name="connectionString">The Azure Service Bus connection string.</param>
+        /// <param name="subscriptionId">The subscription ID where the service bus namespace lives.</param>
+        public Messenger([NotNull]string connectionString, [NotNull]string subscriptionId)
+        {
+            ConnectionString = connectionString;
+            SubscriptionId = subscriptionId;
+        }
+
+        /// <inheritdoc />
+        public async Task Send<T>(T message)
+            where T : class
+        {
+            if (!ServiceBusAdapters.ContainsKey(typeof(T)))
+            {
+                SetupMessageType<T>(10, MessagingTransport.Queue);
+            }
+
+            await ((QueueAdapter<T>)ServiceBusAdapters[typeof(T)]).Send(message).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public async Task Publish<T>(T @event)
+            where T : class
+        {
+            if (!ServiceBusAdapters.ContainsKey(typeof(T)))
+            {
+                SetupMessageType<T>(10, MessagingTransport.Topic);
+            }
+
+            await ((TopicAdapter<T>)ServiceBusAdapters[typeof(T)]).Send(@event).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void Receive<T>(Action<T> callback, int batchSize = 10)
+            where T : class
+        {
+            if (MessageSubs.ContainsKey(typeof(T)))
+            {
+                throw new InvalidOperationException("You already added a callback to this message type. Only one callback per type is supported.");
+            }
+
+            ((QueueAdapter<T>) SetupMessageType<T>(batchSize, MessagingTransport.Queue)).StartReading();
+            MessageSubs.Add(typeof(T), MessagesIn.OfType<T>().Subscribe(callback));
+        }
+
+        /// <inheritdoc />
+        public async Task Subscribe<T>(Action<T> callback, string subscriptionName, int batchSize = 10)
+            where T : class
+        {
+            if (MessageSubs.ContainsKey(typeof(T)))

```

Comment:
locking?

Reply:
Don't think it's really needed since it's a guard method to prevent dup adding callbacks (SetupMessageType has a message lock within it)

But I wondering if we are limiting ourselves with just a single callback per type? that just seems a bit too limiting.

Reply:
You can segregate your callback in any way shape of form, including calling a lot of other things. However, if a developer wants to do something really funky, we don't limit anything, we give direct access to the observable in `GetMessageObservable<T>` and `GetEventObservable<T>`. He then needs to deal with things like keep tracking of subscriptions, etc. but with great power comes great responsibility. We give a very easy to use callback that caters for 95%+ of the cases, when it doesn't, they tap into Rx and there's an API for it.

Reply:
@dguerin @djfr 
yeah I am aware of the lock in the SetupMessageType  but it seems the intention is to prevent the caller from doing so via the exception... which would not always be thrown....

Reply:
@djfr grand, with context this makes sense.